### PR TITLE
Try prevent relative links if URL isn't fully typed out

### LINKED ISSUE #5763: Hyperlinks in blocks create relative links if URL is not fully typed out
## Issue Overview

When adding a hyperlink to a Gutenberg block, if the entire URL is not inputted (In other words, with a "http://www" or a "www"), the link generated is a relative link.

## Steps to Reproduce (for bugs)

1. Create or convert a section of text into a block.
2. Create a hyperlink as usual.
3. However, instead of say, "http://www.google.ca/", make the URL "google.ca"
4. Once the page is published, if you hover over the link, you'll see that it's a relative link, instead of the intended URL.

Using Gutenberg 2.4
Using WordPress 4.9.4
Using Google Chrome Version 65.0.3325.181 (Official Build) (64-bit) on Windows 7

## Expected Behavior

I expect the web link to link to the proper url and to prepend any web address with "http://" if need be.

## Current Behavior

For example, making a hyperlink with "google.ca" links to "http://www.mysite.com/google.ca"

### PR DESCRIPTION
**Pull request by Rahmon**
## Description
How was described on issue #5763: "_When adding a hyperlink to a Gutenberg block, if the entire URL is not inputted (In other words, with a "http://www" or a "www"), the link generated is a relative link_.".

In Classic editor, it's used the [correctURL](https://github.com/WordPress/WordPress/blob/48077799d9d84b41d7ee1ea3a0054f0641813359/wp-includes/js/wplink.js#L87) function that check if link doesn't start with http:, https:, ? # or /, prepend http://.

I try do the same in Gutenberg

Closes #5763  

## How Has This Been Tested?
This has been tested with "npm test" and manually on Chrome and Firefox.

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

### CODE REVIEW

File: blocks/rich-text/format-toolbar/index.js [hunk 1]
```diff
@@ -110,9 +110,22 @@ class FormatToolbar extends Component {
 		this.setState( { isEditingLink: false, isAddingLink: true, newLinkValue: this.props.formats.link.value } );
 	}
 
+	correctLink( link ) {
```
Comment (mtias) [inline blocks/rich-text/format-toolbar/index.js]: Such a function should probably live in https://github.com/WordPress/gutenberg/tree/master/utils

Comment (Rahmon) [reply blocks/rich-text/format-toolbar/index.js]: Do I need create a new file, right?

File: utils/hyperlink.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Utility used to check if URL doesn't start with http:, http:, ? # or /, prepend http://
+ *
+ * @param {string} link Hyperlink to correct.
+ * @return {string} Hyperlink corrected.
+ */
+
```
Comment (aduth) [inline utils/hyperlink.js]: With the newline here, I don't know that JSDoc parsers would associate with the function. Suggest removal.

File: utils/hyperlink.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+/**
```
Comment (aduth) [inline utils/hyperlink.js]: Seems like a good candidate for unit testing various behaviors.

**Review by aduth — COMMENTED** [review]
`correctLink` is a vague name. It's not to me what an "incorrect" link is by the name of the function alone.

Further, I'd kindly disagree with Matias here, mostly in that a `@wordpress/url` module already exists:

- https://www.npmjs.com/package/@wordpress/url
- https://github.com/WordPress/packages

The `./utils` directory is... an unfortunate existence.

Thoughts on a clearer name for the function:

- `isRelative` (prepend protocol from consumer if `true`)
- `addDefaultProtocol`
- `prependHTTP`

Aside: I landed at the last of these as pursuit of understanding what the function is doing led me to wonder if it's already a solve problem, and it is: https://www.npmjs.com/package/prepend-http . Whether we care to bring in a new package for what is arguably a small and isolated problem is, well... the same discussion we should be having about whether to pull it out into a utilities file at all, or just keep the logic inline :smile:

File: utils/hyperlink.js (hunk 3 above)
Comment (Rahmon) [reply utils/hyperlink.js]: I thought the same while coding.. xD

**Comment by Rahmon** [comment]
@aduth 

I agree with you, `correctLink` isn't a good name but naming things is a hard thing..rs. 

I'm considering put this function in `@wordpress/url` module because make more sense for me now.

**Comment by pento** [comment]
Note: I've opened [a PR](https://github.com/WordPress/packages/pull/112) to add this functionality to the `@wordpress/url` package.

**Comment by Rahmon** [comment]
@pento great. I'll wait for merge to update this PR.

**Comment by pento** [comment]
@Rahmon: Awesome! `@wordpress/url` 1.1.0 is available, you can make use of that now. (Feel free to create new PR if you don't feel like re-using this one.) 🙂

**Review by aduth — APPROVED** [review]
Looks good 👍 

Since the dependency version was updated, we needed to update `package-lock.json` as well. This is done automatically when updating via the `npm` command-line tool, so I'm assuming the change was made manually to the `package.json` file.

**Pull request merged**